### PR TITLE
Wal back pressure v2

### PR DESCRIPTION
**Pull request by hlinnaka**
There's a PR for WAL back pressure in the 'zenith', repo, but I'm creating this second PR in the 'postgres' repo to have a place for commenting on the postgres changes.

### CODE REVIEW

File: src/backend/storage/freespace/freespace.c [hunk 1]
```diff
@@ -637,10 +637,13 @@ fsm_extend(Relation rel, BlockNumber fsm_nblocks)
 
 	while (fsm_nblocks_now < fsm_nblocks)
 	{
-		PageSetChecksumInplace((Page) pg.data, fsm_nblocks_now);
-
-		smgrextend(rel->rd_smgr, FSM_FORKNUM, fsm_nblocks_now,
-				   pg.data, false);
+		/* ZENITH: Initialize FSM pages through buffer cache to prevent loading them from pageserver */
```
Comment (hlinnaka) [inline src/backend/storage/freespace/freespace.c:640]: Why is this change needed? Is it still needed with the PD_WAL_LOGGED removal?

Comment (ololobus) [reply src/backend/storage/freespace/freespace.c:640]: Hm, @knizhnik, am I correct that with `smgrextend` this will hit `zenith_wallog_page` and freshly allocated FSM page will be WAL logged? And this doesn't have much sense, so this is a performance optimization, but is it related to the back pressure?

File: src/backend/utils/misc/guc.c [hunk 2]
```diff
@@ -2898,6 +2898,30 @@ static struct config_int ConfigureNamesInt[] =
...
 
+	{
+		{"max_replication_write_lag", PGC_POSTMASTER, REPLICATION_SENDING,
+			gettext_noop("Maximal write lag between master and replicas."),
+			gettext_noop("When lag between minimal write position of replica and current LSN exeeds this value,"
+						 "backends are blocked"),
+			GUC_UNIT_MB,
+		},
+		&max_replication_write_lag,
+		1024, 0, INT_MAX, /* it should not be smaller than maximal size of WAL record */
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply src/backend/utils/misc/guc.c:2909]: Yes, with large WAL records we get a trouble here. But it is not only back pressure specific problem. Our current approach with not processing incomplete WAL records can suffer from arbitrary large WAL records. I do not have any good idea at this moment how to handle this problem. But frankly speaking I do not think that there will be some Zenith customers, using logical replication and especially with custom WAL records. 

Please also notice that currently we use default replication level=`replica`, not `logical`. It cause quite noticeable reduce of WAL size. So right now logical replication is not supported at all. And it is not clear for me whether we may let allow users to switch it on.

**Comment by ololobus** [comment]
@knizhnik, I've went ahead and rebased the branch.

- Added/changed some comments
- Introduced `UnknownXLogRecPtr`. First tried to replace it with `InvalidXLogRecPtr`, but this uint64 max indeed simplifies comparisons a lot, so let's live with it for some time
- Moved this FSM optimization into a separate commit for a proper history and did the same for VM

If you and everyone else are OK with it, I'll squash commits and make one for back pressure and one for FSM/VM optimization

**Comment by ololobus** [comment]
OK, it was a bad idea to add log message with INFO level. Many regression tests fail with it:
```log
--- /home/circleci/project/vendor/postgres/src/test/regress/expected/transactions.out	2021-10-20 10:28:52.515876675 +0000
+++ /tmp/test_output/test_pg_regress/regress/results/transactions.out	2021-10-20 10:29:51.268316644 +0000
@@ -658,7 +658,9 @@
 ERROR:  portal "ctt" cannot be run
 COMMIT;
 DROP FUNCTION create_temp_tab();
+INFO:  slept for 100 ms while waiting for all replicas to catch up
 DROP FUNCTION invert(x float8);
```
